frontend/public/module/ws-factory.ts

```diff
@@ -0,0 +1,344 @@
+const createURL = (host: string, path: string): string => {
+  let url;
+
+  if (host === 'auto') {
+    if (location.protocol === 'https:') {
+      url = 'wss://';
+    } else {
+      url = 'ws://';
+    }
+    url += location.host;
+  } else {
+    url = host;
+  }
+
+  if (path) {
+    url += path;
+  }
+
+  return url;
+};
+
+export type WSOptions = {
+  host: string;
+  path: string;
+  subprotocols: string[];
+  reconnect?: boolean;
+  jsonParse?: boolean;
+  bufferMax?: number;
+  bufferFlushInterval?: number;
+  timeout?: number;
+};
+
+type GenericHandler<T = any> = (data: T) => void;
+type OpenHandler = GenericHandler<undefined>; // nothing is sent
+type CloseHandler = GenericHandler<CloseEvent>;
+type ErrorHandler = GenericHandler<Event>;
+/**
+ * The WebSocket can send JSON that is parsed, or we just send it through as-is
+ */
+type MessageDataType = object | any;
+type MessageHandler = GenericHandler<MessageDataType>;
+type DestroyHandler = GenericHandler<undefined>;
+type BulkMessageHandler = GenericHandler<MessageDataType>;
+
+type RegisterHandler =
```

Comment:
Not sure this guy is necessary. The `GenericHandler` is the super-type for all our handlers... could have just used that.